lua/plenary/async_lib/async.lua

```diff
@@ -0,0 +1,190 @@
+local co = coroutine
+local uv = vim.loop
+
+local M = {}
+
+local Future = {}
+
+function Future.new(opts)
+  vim.validate {
+    leaf = {opts.leaf, 'boolean'},
+    args = {opts.args, 'table'},
+    func = {opts.func, 'function'},
+  }
+
+  if opts.leaf then
+    table.insert(opts.args, function() end)
```

Comment:
I wouldn't recommend appending the callback onto arguments like this. I added a similar (but less sophisticated) async library to [gitsigns.nvim](http://github.com/lewis6991/gitsigns.nvim)  and I found a nasty side effect (I would even call it a bug).

Basically the issue arises when the final argument to an async function is `nil`. Because of how lua handles nil in tables, a value appended to the table will take the final arguments place

```
local function wrap(f)
  return function(...)
    local params = {...}
    table.insert(params, 'hello')
    f(table.unpack(params))
  end
end

local foo = wrap(function(a, b, c)
  print('a='..a)
  print('b='..b)
  print('c='..c)
end)

foo('A', 'B', nil)
```

Outputs:
```
a=A
b=B
c=hello
```

Because `c=nil`, the appended callback takes c's place so the arg list becomes `{ 'A', 'B', 'hello' }`. So now when `c` is accessed in `foo`, you end up accessing the async callback.

This lead me to rewriting the async lib in https://github.com/lewis6991/gitsigns.nvim/commit/0e12a472151aab4187f429479e1b4b4212157312 so all async functions are required to have the signature `function(args...) -> function(callback)`. That way it is impossible for nil's to creep in a shift the arguments.

Looking at line 61, my guess is that this implementation is susceptible to this bug too.

Reply:
It seems like you are looking at an outdated version. async functions already have the `function(args...) -> function(callback)` signature. However, this is still an issue for leaf futures, (aka futures that are create from `a.wrap`ed functions). This is impossible to solve however because they are leaf futures. Furthermore, `a.wrap` is mostly used for libuv functions that take a callback, and you shouldn't be passing in nil anyway as those functions must take all parameters. If you want to accept `nil` then you can just create a `async` wrapper around the base leaf future.

Reply:
>  If you want to accept nil then you can just create a async wrapper around the base leaf future.

For me the issue wasn't about needing to accept nil values, but nil values __accidentally__ being passed to functions which caused things to blow up with very unhelpful error messages and was incredibly tedious to debug with the eventual issue being quite simple.

Reply:
`nil` is supposed to be idiomatic :man_facepalming:. thats lua for you

There is just no way to solve that though. The only way is to encase the base leaf future like this:
```lua
local validated_fs_open = async(function(path, flag, mode)
   assert(mode, "mode must not be nil")
   assert(flag, "flag must not be nil")
   assert(path, "path must not be nil")
   return await(a.uv.fs_open(path, flag, mode))
end)
```
we could also have wrap take the number of arguments that it should expect a return a wrapped function that validates the arguments. for example:

```lua
-- use with CPS function, creates future factory
-- optional argc will validate argument count
M.wrap = function(func, argc)
  assert(type(func) == "function", "type error :: expected func, got " .. type(func))

  return function(...)
    local params = {...}

    if argc then
      assert(#params == argc, "Not enough params")
    end

    local function future(step)
      if step then
        table.insert(params, step)
        return func(unpack(params))
      else
        return co.yield(future)
      end
    end
    return future
  end
end
```
actually that looks pretty good, what do you think about that?

Reply:
The second solution sounds pretty good to me. The first version of gitsigns actually used the `debug` module to get the number of arguments and that's how I initially solved the nil issue. Apparently though the `debug` module is slow so a rewrote the lib to avoid using it.

But yes, having `wrap` take the number of arguments sounds good to me.

Reply:
Also if you have `argc` then you just need to insert `step` at the right position, I don't think you need to assert `#params == argc`, that actually prevents you from using `nil`.

Reply:
I don't like using `nil` though for a leaf future. `a.wrap` is mostly used for wrapping libuv functions and some low level utilities like `Condvar` and the like which do not and should not accept nil. If nil is needed then just create a non-leaf future by using `async`

Reply:
Well as long as it's properly documented I guess that's ok. I would have thought that since it's apart of the external API, we should make it handle `nil` as best as possible. If you're only using it internally for other things like libuv functions then that's different.

Like you said `nil` is idiomatic 😆.

Reply:
I don't think we should assume that libuv functions are the only ones that will be wrapped. a plugin dev might want to use some other lib that takes `nil` arg before callback.

Reply:
so should `argc` then be required, because right now it is optional

Reply:
yes it is a bit unfortunate but the only way to do it safely.

Reply:
I think it's ok for it to be optional, as long as people like me have a path to avoiding that horrible issue. But I don't mind either way.

Reply:
ok im going to make it *not* optional once I change all the wrapped libuv functions to have the `argc`.